colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -557,55 +596,75 @@ def schedule_f(
                 input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
 
         # Here, let input_obj.requires_grad_()
-        tree_map(torch.Tensor.requires_grad_, input_obj)
+        # if input_obj is not None:
+        if not isinstance(input_obj, torch.Tensor):
+            tree_map(require_grad, input_obj)
+
+        # Also requires_grad_ for micro_batch in stage 0 chunk 0 fwd,
+        # tree_map(torch.Tensor.requires_grad_, micro_batch)
 
         # Step2: fwd step
         output_obj = self.forward_step(
             model_chunk=model_chunk,
             model_chunk_id=model_chunk_id,
+            micro_batch=micro_batch,
             input_obj=input_obj,
             criterion=criterion,
             accum_loss=accum_loss,
             outputs=outputs,
         )
+
+        # Step3:
+        # 3-1:detach output; detach output for send fwd;
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # We should not detach bwd LOSS
             pass
         else:
-            detached_output_obj = output_obj.clone().detach()
+            # detach output
+            detached_output_obj = tree_map(detach, output_obj)
+        # 3-2 clone output
+        output_obj = tree_map(clone, output_obj)
```

Comment:
we should clone detached output obj

Reply:
Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/7e6f793c5182d7da95e443967be0a6c9777bd01e;